CASMCMS-7327: Add information on the flow of CFS sessions (#1395)
Cray-HPE/docs-csm: operations/configuration_management/CFS_Flow_Diagrams.md

# CFS Flow

> 

Reviewer: ```suggestion

* [Single Session Flow](#single_session_flow)
* [Automated Session Flow](#automated_session_flow)

```

> ## Single Session Flow

Reviewer: ```suggestion
<a name="single_session_flow"></a>
## Single Session Flow
```

This section covers the components and actions taken when a user or service creates a session using the CFS sessions endpoint.

![Single Session Flow](../../img/operations/CFS_Single_Session_Flow.png)

1. A user creates a CFS configuration.
1. A user creates a CFS session, causing a session record to be created.
1. When a session record is created, the CFS-API also posts an event to a Kafka queue.
1. The CFS-Operator is always monitoring the Kafka queue, and handles events as they come in.
1. The CFS-Operator creates a Kubernetes job for the session in response to a session creation event.
1. CFS-Operator monitors the Kubernetes job (not the pod) for completion.
1. When the Kubernetes job is complete, CFS-Operator updates the session record in the CFS-API.

> ## Automated Session Flow

Reviewer: ```suggestion
<a name="automated_session_flow"></a>
## Automated Session Flow
```

This section covers the components and actions taken when a user or service sets a desired configuration for a component, causing automatic configuration to occur.

![Automated Session Flow](../../img/operations/CFS_Automated_Session_Flow.png)

1. A user creates a CFS configuration.

> 1. A user sets the desired configuration for some number of components. This may immediately trigger configuration if the component is enabled in CFS and the desired configuration not yet been applied. If so, skip to step 4.

Reviewer: ```suggestion
1. A user sets the desired configuration for some number of components. This may immediately trigger configuration if the component is enabled in CFS and the desired configuration not yet been applied. If that is the case, then skip to [CFS-Batcher creates sessions](#cfs_batcher_creates_sessions).
```

1. When a node reboots, the CFS-State-Reporter runs on the node, and contacts the CFS-API to enable and clear the state for the current node. This will always trigger the next steps for configuration as long as a desired configuration is set for the node.

> 1. The CFS-Batcher monitors the CFS-API, periodically querying for enabled components with a "pending" configuration status.  (Status is determined by the API at query time based on the desired and current state of the component).  Any components found are placed into batches.

Reviewer: ```suggestion
1. <a name="cfs_batcher_monitors_for_pending"></a>The CFS-Batcher monitors the CFS-API, periodically querying for enabled components with a `pending` configuration status. (The status is determined by the API at query time based on the desired and current state of the component). Any components found are placed into batches.
```

> 1. The CFS-Batcher calls the CFS-API to create sessions for each batch, using the "ansible-limit" parameter to limit each session to the components in a batch.

Reviewer: ```suggestion
1. <a name="cfs_batcher_creates_sessions"></a>The CFS-Batcher calls the CFS-API to create sessions for each batch, using the `ansible-limit` parameter to limit each session to the components in a batch.
```

> 1. The CFS-API creates a session record and session creation event. (see Single Session Flow) 

Reviewer: ```suggestion
1. The CFS-API creates a session record and session creation event. See [Single Session Flow](#single_session_flow) for details.
```

> 1. The CFS-Operator monitors for session creation events. (see Single Session Flow) 

Reviewer: ```suggestion
1. The CFS-Operator monitors for session creation events. See [Single Session Flow](#single_session_flow) for details.
```

> 1. The CFS-Operator creates a Kubernetes job. (see Single Session Flow)

Reviewer: ```suggestion
1. The CFS-Operator creates a Kubernetes job. See [Single Session Flow](#single_session_flow) for details.
```

> 1. The Ansible Execution Environment contains a custom Ansible callback plugin that calls CFS to update the status of all components effected after each play.

Reviewer: ```suggestion
1. The Ansible Execution Environment contains a custom Ansible callback plugin that calls CFS to update the status of all components affected after each play.
```

> 1. If the component does not have a "configured" status due to failure or other reasons, return to step 4.

Reviewer: ```suggestion
1. If the component does not have a `configured` status due to failure or other reasons, return to [CFS-Batcher monitors for enabled `pending` components](#cfs_batcher_monitors_for_pending).
```
